python/ray/workflow/tests/test_wait.py

```diff
@@ -0,0 +1,150 @@
+from ray.tests.conftest import *  # noqa
+
+import time
+import pytest
+import ray
+from ray import workflow
+from ray.workflow.common import Workflow
+from ray.workflow.tests import utils
+from ray.exceptions import RaySystemError
+
+
+@workflow.step
+def sleep_identity(x: int):
+    time.sleep(x)
+    return x
+
+
+@workflow.step
+def wait_simple():
+    ws = [
+        sleep_identity.step(1),
+        sleep_identity.step(3),
+        sleep_identity.step(10),
+        sleep_identity.step(2),
+        sleep_identity.step(12),
+    ]
+    return workflow.wait(ws, num_returns=4, timeout=5)
+
+
+@workflow.step
+def get_all(ready, unready):
+    return ready, unready
+
+
+@workflow.step
+def filter_all(wait_results):
+    ready, unready = wait_results
+    return get_all.step(ready, unready)
+
+
+@workflow.step
+def composite():
+    w = wait_simple.step()
+    return filter_all.step(w)
+
+
+@workflow.step
+def recursive_wait(s):
+    ready, unready = s
+    if len(unready) == 2 and not isinstance(unready[0], Workflow):
+        ready_2, unready = unready
+        print(ready, (ready_2, unready))
+        ready += ready_2
+
+    if not unready:
+        return ready
+    w = workflow.wait(unready)
+    return recursive_wait.step([ready, w])
+
+
+@pytest.mark.parametrize(
+    "workflow_start_regular",
+    [{
+        "num_cpus": 8
+        # We need more CPUs, otherwise task execution could be blocked.
+    }],
+    indirect=True)
+def test_wait_basics(workflow_start_regular):
+    ready, unready = wait_simple.step().run()
+    assert ready == [1, 3, 2]
+
+    ready, unready = composite.step().run()
+    assert ready == [1, 3, 2]
+    assert unready == [10, 12]
+
+    w = wait_simple.step()
+    assert recursive_wait.step(w).run() == [1, 3, 2, 10, 12]
+
+    with pytest.raises(TypeError):
+        workflow.wait([1, 2])
+
+
+@workflow.step
+def do_task(i):
+    import random
+    time.sleep(random.random())
+    return "task {}".format(i)
+
+
+@workflow.step
+def report_results(wait_result):
+    ready, remaining = wait_result
+    for result in ready:
+        print("Completed", result)
+    if not remaining:
+        return "All done"
+    else:
+        return report_results.step(
+            workflow.wait(remaining, num_returns=max(len(remaining) // 2, 1)))
+
+
+@pytest.mark.parametrize(
+    "workflow_start_regular",
+    [{
+        "num_cpus": 8
+        # We need more CPUs, otherwise task execution could be blocked.
+    }],
+    indirect=True)
+def test_wait_example(workflow_start_regular):
+    tasks = [do_task.step(i) for i in range(20)]
+    report_results.step(workflow.wait(tasks)).run()
+
+
+@workflow.step
+def filter_all_2(wait_results):
+    assert utils.check_global_mark()
```

Comment:
Should we add some kind of assertion that the `ready` results that we get before and after failure are the same?